vowpalwabbit/parse_example.cc

```diff
@@ -50,29 +50,113 @@ int read_features_string(vw* all, v_array<example*>& examples)
   return static_cast<int>(num_chars_initial);
 }
 
+class feature_group_helper
+{
+public:
+  using char_map = std::array<unsigned char, NUM_NAMESPACES>;
+  using affix_map = std::array<uint64_t, NUM_NAMESPACES>;
+  using bool_map = std::array<bool, NUM_NAMESPACES>;
+  using namespace_dictionaries = std::vector<std::shared_ptr<feature_dict>>;
+  using dictionaries_map = std::array<namespace_dictionaries, NUM_NAMESPACES>;
+
+  feature_group_helper(example* current_example, bool redefine_char_flag, char_map* redefine_char_map,
+      affix_map* affix_features, bool_map* spelling_features, dictionaries_map* namespace_dictionaries_map,
+      uint32_t hash_seed, hash_func_t hash_fn)
+      : _feature_group_char(0)
+      , _hash(0)
+      , _hash_seed(hash_seed)
+      , _p_current_feature_group(nullptr)
+      , _p_affix_feature_group(nullptr)
+      , _p_spelling_feature_group(nullptr)
+      , _p_dictionary_feature_group(nullptr)
+      , _current_example(current_example)
+      , _redefine_char_flag(redefine_char_flag)
+      , _redefine_char_map(redefine_char_map)
+      , _namespace_dictionaries_map(namespace_dictionaries_map)
+      , _affix_features(affix_features)
+      , _spelling_features(spelling_features)
+      , _hash_function(hash_fn)
+  {
+  }
+  unsigned char feature_group_char() const { return _feature_group_char; }
+  uint64_t feature_group_hash() const { return _hash; }
+  features& get_current_fg() { return get_feature_group(_p_current_feature_group, feature_group_char()); }
+  features& get_affix_fg() { return get_feature_group(_p_affix_feature_group, affix_namespace); }
+  features& get_spelling_fg() { return get_feature_group(_p_spelling_feature_group, spelling_namespace); }
+  features& get_dictionary_fg() { return get_feature_group(_p_dictionary_feature_group, dictionary_namespace); }
+  void set_current_fg(const VW::string_view fg_name)
+  {
+    _feature_group_name = fg_name;
+    auto namespace_char = static_cast<unsigned char>(_feature_group_name[0]);
+    if (_redefine_char_flag) namespace_char = (*_redefine_char_map)[namespace_char];  // redefine feature_group_char
+    _feature_group_char = namespace_char;
+    _hash = _hash_function(_feature_group_name.begin(), _feature_group_name.length(), _hash_seed);
+    _p_current_feature_group = nullptr;  // reset cached feature group
+    _affix = (*_affix_features)[_feature_group_char];
+    _spelling = (*_spelling_features)[_feature_group_char];
+    _namespace_dictionaries = &(*_namespace_dictionaries_map)[feature_group_char()];
```

Comment:
nit: does this need to be dereferenced and then referenced? Can we just do:

`_namespace_dictionaries = _namespace_dictionaries_map`?

Reply:
I don't see how.  The [] operator returns a reference of the value stored in the container.  & is used to get the address to that reference.  Maybe parentheses can make things more clear.

Reply:
oh, nevemind, I didn't see the [] operator, github diff put it on a different line